packages/transactions/src/new-signatures.ts

```diff
@@ -0,0 +1,96 @@
+import { Address, getAddressFromPublicKey } from '@solana/addresses';
+import { Decoder } from '@solana/codecs-core';
+import { getBase58Decoder } from '@solana/codecs-strings';
+import {
+    SOLANA_ERROR__TRANSACTION__FEE_PAYER_SIGNATURE_MISSING,
+    SOLANA_ERROR__TRANSACTION__SIGNATURES_MISSING,
+    SolanaError,
+} from '@solana/errors';
+import { Signature, SignatureBytes, signBytes } from '@solana/keys';
+
+import { CompilableTransaction } from './compilable-transaction';
+import { compileMessage } from './message';
+import { OrderedMap } from './ordered-map';
+import { getCompiledMessageEncoder } from './serializers';
+
+export type TransactionMessageBytes = Uint8Array & { readonly __brand: unique symbol };
+
+export interface INewTransactionWithSignatures {
+    messageBytes: TransactionMessageBytes;
+    signatures: OrderedMap<Address, SignatureBytes | null>;
+}
+
+export interface INewFullySignedTransaction extends INewTransactionWithSignatures {
+    readonly __brand: unique symbol;
+}
+
+let base58Decoder: Decoder<string> | undefined;
+
+export function newGetSignatureFromTransaction(transaction: INewTransactionWithSignatures): Signature {
+    if (!base58Decoder) base58Decoder = getBase58Decoder();
+
+    // We have ordered signatures from the compiled message accounts
+    // first signature is the fee payer
+    const signatureBytes = transaction.signatures.firstValue();
+    if (!signatureBytes) {
+        throw new SolanaError(SOLANA_ERROR__TRANSACTION__FEE_PAYER_SIGNATURE_MISSING);
+    }
+    const transactionSignature = base58Decoder.decode(signatureBytes);
+    return transactionSignature as Signature;
+}
+
+export async function newPartiallySignTransaction<TTransaction extends CompilableTransaction>(
```

Comment:
To clarify, this would be an overload with the type `(CryptoKeyPair[], ITransactionWithSignatures) -> ITransactionWithSignatures`, used to add additional signers to a partially signed tx by signing the same message bytes?

Reply:
My understanding here is that the new `partialSignTransaction` should either:

- Accept a `TransactionMessage` and return a new `Transaction`, or
- Accept a `TTransaction extends Transaction` and return `TTransaction`.

And the new `signTransaction` should either:

- Accept a `TransactionMessage` and return a new `Transaction & FullySignedTransaction`, or
- Accept a `TTransaction extends Transaction` and return `TTransaction & FullySignedTransaction`.

_Note that I'm naming types based on my previous comment._

Wdyt?

Reply:
I've refactored this:
- partial sign is `NewTransaction -> NewTransaction`
- sign is `NewTransaction -> NewFullySignedTransaction`

The logic to compile from `TransactionMessage -> NewTransaction` is further up the stack. I think it makes sense to just keep these steps separate because this ends up doing too much otherwise and using completely different types. It'll be trivial to write a helper to compile + sign if someone wants to.

I also didn't use `TTransaction extends Transaction` because I don't think anyone will need to extend `NewTransaction`. It's just message + signature, you really can't/shouldn't do anything else with it.